File: solutions/javascript/list-ops/2/list-ops.js
export class List {
  constructor(list) {
    this.values = list || []
  }

  myPush(list, el) {
    return [...list, el]
  }

  foldLeft(f, start, [head, ...tail]) {
    return head ? this.foldLeft(f, f(start, head), tail) : start
  }
  
  foldRight(f, start, [head, ...tail]) {
    return head ? f(this.foldRight(f, start, tail), head) : start
  }

  append(otherList) {
    return new List(this.foldLeft(
      (s, h) => this.myPush(s, h),
      this.values,
      otherList.values
    ))
  }

  concat(list) {
    return new List(this.foldLeft(
      (s, h) => [...s, ...h.values],
      this.values,
      list.values
    ))
  }

  filter(f) {
    return new List(this.foldLeft(
      (s, h) => f(h) ? this.myPush(s, h) : s,
      [],
      this.values
    ))
  }

  map(f) {
    return new List(this.foldLeft(
      (s, h) => this.myPush(s, f(h)),
      [],
      this.values
    ))
  }

  length() {
    return this.foldLeft(
      (s, h) => 1 + s,
      0,
      this.values
    )
  }

  foldl(f, start) {
    return this.foldLeft(f, start, this.values)
  }

  foldr(f, start) {
    return this.foldRight(f, start, this.values)
  }

  reverse() {
    return new List(
      this.foldRight(
        (s, h) => this.myPush(s, h),
        [],
        this.values
      )
    )
  }
}

File: solutions/go/list-ops/2/list_ops.go
package listops

// IntList is an abstraction of a list of integers which we can define methods on
type IntList []int

func (s IntList) Foldl(fn func(int, int) int, initial int) int {
	if s.Length() == 0 {
		return initial
	}

	acc := initial

	for _, val := range(s) {
		acc = fn(acc, val)
	}

	return acc
}

func (s IntList) Foldr(fn func(int, int) int, initial int) int {
	if s.Length() == 0 {
		return initial
	}

	acc := initial

	for i := s.Length() - 1; i >= 0; i-- {
		acc = fn(s[i], acc)
	}

	return acc
}

func (s IntList) Filter(fn func(int) bool) IntList {
	res := IntList{}

	for _, val := range(s) {
		if fn(val) {
			res = res.Append([]int{val})
		}
	}

	return res
}

func (s IntList) Length() int {
	length := 0

	for range(s) {
		length++
	}

	return length
}

func (s IntList) Map(fn func(int) int) IntList {
	res := IntList{}

	for _, val := range(s) {
		res = res.Append(IntList{fn(val)})
	}

	return res
}

func (s IntList) Reverse() IntList {
	res := IntList{}

	for i := s.Length() - 1 ; i >= 0; i-- {
		res = res.Append(IntList{s[i]})
	}

	return res
}

func (s IntList) Append(lst IntList) IntList {
	appended := make(IntList, s.Length()+lst.Length())

	copy(appended, s)
	copy(appended[s.Length():], lst)

	return appended
}

func (s IntList) Concat(lists []IntList) IntList {
	for _, list := range(lists) {
		s = s.Append(list)
	}

	return s
}
